File: 0901-1000/0998-Maximum Binary Tree II/0998-Maximum Binary Tree II.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def insertIntoMaxTree(self, root: TreeNode, val: int) -> TreeNode:
        if root is None or root.val < val:
            node = TreeNode(val)
            node.left = root
            return node
        else:
            root.right = self.insertIntoMaxTree(root.right, val)
            return root


class Solution2:
    def insertIntoMaxTree(self, root: TreeNode, val: int) -> TreeNode:
        prev = None
        curr = root
        while curr and curr.val > val:
            prev = curr
            curr = curr.right
        
        node = TreeNode(val)
        node.left = curr
        if prev:
            prev.right = node
        
        return root if prev else node
   

File: 0901-1000/0998-Maximum Binary Tree II/0998-Maximum Binary Tree II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoMaxTree(TreeNode* root, int val) {
        if (root == nullptr || root->val < val) {
            TreeNode* node = new TreeNode(val);
            node->left = root;
            return node;
        }
        else {
            root->right = insertIntoMaxTree(root->right, val);
            return root;
        }
    }
};


class Solution {
public:
    TreeNode* insertIntoMaxTree(TreeNode* root, int val) {
        TreeNode *prev = nullptr, *curr = root;
        while (curr && curr->val > val) {
            prev = curr;
            curr = curr->right;
        }
        
        TreeNode* node = new TreeNode(val);
        node->left = curr;
        if (prev) {
            prev->right = node;
        }
        
        return prev ? root : node;
    }
};
